utils/frame/remote-externalities/src/rpc_api.rs

```diff
@@ -18,36 +18,65 @@
 //! WS RPC API for one off RPC calls to a substrate node.
 // TODO: Consolidate one off RPC calls https://github.com/paritytech/substrate/issues/8988
 
-use super::*;
+use sp_runtime::{generic::SignedBlock, traits::{Block as BlockT, Header as HeaderT}};
+use jsonrpsee_ws_client::{WsClientBuilder, WsClient, v2::params::JsonRpcParams, traits::Client};
 
 /// Get the header of the block identified by `at`
-pub async fn get_header<B: BlockT, S: AsRef<str>>(from: S, at: B::Hash) -> Result<B::Header, String>
+pub async fn get_header<Block, S>(from: S, at: Block::Hash) -> Result<Block::Header, String>
 where
-	B::Header: serde::de::DeserializeOwned,
+	Block: BlockT,
+	Block::Header: serde::de::DeserializeOwned,
+	S: AsRef<str>,
 {
-	use jsonrpsee_ws_client::traits::Client;
-	let at = serde_json::to_value(at)
-		.map_err(|e| format!("Block hash could not be converted to JSON due to {:?}", e))?;
-	let params = vec![at];
-	let client = WsClientBuilder::default()
-		.max_request_body_size(u32::MAX)
-		.build(from.as_ref())
-		.await
-		.map_err(|e| format!("`WsClientBuilder` failed to build do to {:?}", e))?;
-	client.request::<B::Header>("chain_getHeader", JsonRpcParams::Array(params))
+	let params = vec![hash_to_json::<Block>(at)?];
+	let client = build_client(from).await?;
+
+	client.request::<Block::Header>("chain_getHeader", JsonRpcParams::Array(params))
 		.await
 		.map_err(|e| format!("chain_getHeader request failed due to {:?}", e))
 }
 
 /// Get the finalized head
-pub async fn get_finalized_head<B: BlockT, S: AsRef<str>>(from: S) -> Result<B::Hash, String> {
-	use jsonrpsee_ws_client::traits::Client;
-	let client = WsClientBuilder::default()
+pub async fn get_finalized_head<Block, S>(from: S) -> Result<Block::Hash, String>
+where
+	Block: BlockT,
+	S: AsRef<str>,
+{
+	let client = build_client(from).await?;
+
+	client.request::<Block::Hash>("chain_getFinalizedHead", JsonRpcParams::NoParams)
+		.await
+		.map_err(|e| format!("chain_getFinalizedHead request failed due to {:?}", e))
```

Comment:
Not sure how deeply these errors will chain, but in case the chains get deep, it's easier to parse if we use a punctuation delimiter instead of the text "due to".

```suggestion
		.map_err(|e| format!("chain_getFinalizedHead request failed: {:?}", e))
```